### ISSUE TITLE
Examples with nice error messages

### ISSUE THREAD

**Issue by adamchalmers** [winnow-rs/winnow#337]
### Please complete the following tasks

- [X] I have searched the [discussions](https://github.com/winnow-rs/winnow/discussions)
- [X] I have searched the [open](https://github.com/winnow-rs/winnow/issues) and [rejected](https://github.com/winnow-rs/winnow/issues?q=is%3Aissue+label%3AS-wont-fix+is%3Aclosed) issues

### winnow version

0.5.15

### Describe your use case

I am writing a language parser with Winnow, and I'd like to get nice, user-friendly errors that show exactly where in their source code parsing failed, and why the parser failed.

Nom's documentation shows a good example of this: https://github.com/rust-bakery/nom/blob/main/doc/error_management.md, using `nom::error::convert_error` to get a good human-readable error message.

I couldn't find a similar function for winnow. Ideally such a function would be described in the [error management tutorial](https://docs.rs/winnow/latest/winnow/_tutorial/chapter_7/index.html). 

Basically I can't figure out how I'm supposed to create a nice, human-readable error message for consumers of my Rust binary from the current Winnow error types.

### Describe the solution you'd like

Update the error management docs to show how to produce human-readable errors. Something like the Nom docs where they say "But by looking at the original input and the chain of errors, we can build a more user friendly error message. The `nom::error::convert_error` function can build such a message" would be ideal.

### Alternatives, if applicable

If you don't want to include a function like `convert_error` in Winnow, I understand. But in that case, providing an example implementation in the error management docs, to show programmers how to create their own, would be helpful.

**Comment by epage** [comment]
Could you give an example piece of code, what error you get, and what you'd expect for the error?

Depending on what you do, the errors are of the level used by `toml_edit` / `toml` which cargo exposes.

**Comment by adamchalmers** [comment]
If I edit [Winnow's JSON example https://github.com/winnow-rs/winnow/blob/552e417f32f27f16b22af38398151cf98d94529c/examples/json/main.rs#L20] and change the `"a"` to `"a`, then the example outputs

```
ParseError {
    input: "  { \"a\t: 42,\n  \"b\": [ \"x\", \"y\", 12 ] ,\n  \"c\": { \"hello\" : \"world\"\n  }\n  } ",
    offset: 16,
    inner: Verify,
}
```

I'd like to get errors like

```
at line 0, in json:
   { "a : 42,
       ^
 expected '"', found ' '
```

**Comment by adamchalmers** [comment]
Ah, great, that results in much nicer output. I couldn't figure out how to get this nice output from reading the API docs on docs.rs but I guess the examples/ subdirectory is better for this. 

I'm currently busy writing my actual parser with Winnow, just getting the happy path to work, but when I'm ready to look at user-friendly errors, I'll return to this issue and update the API docs so it's clearer how to produce nice error messages.

**Comment by vc1** [comment]
get error offset and custom message

<img width="908" alt="image" src="https://github.com/winnow-rs/winnow/assets/9018903/190d343f-d259-4751-9336-2fe482a8d411">

<img width="740" alt="image" src="https://github.com/winnow-rs/winnow/assets/9018903/1a5542d7-7fb1-4aec-803c-dabbaabc69bd">

**Comment by adamchalmers** [comment]
OK, so upon closely rereading the docs, I've realized one big difference between [`ParseError`](https://docs.rs/winnow/latest/winnow/error/struct.ParseError.html) and [`ContextError`](https://docs.rs/winnow/latest/winnow/error/struct.ContextError.html) is that

* `ParseError` has more information (i.e. the offset within the input at which the error occurred), but can only be returned from `.parse`, _not_ `.parse_next`
* `ContextError` has less information and comes from `.parse_next`

I personally found this confusing so I think I'll open a PR to make this comparison clearer in the docs.

**Comment by adamchalmers** [comment]
I'd also like to add examples for how to handle errors with custom tokens. The built-in pretty errors when your input type is &str are great. But I think an example for custom tokens, e.g. &[T] would be helpful. How do y'all feel about something like this example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=fc3d358804257748e1f6fe09e5d1e836

**Comment by epage** [comment]
> On a similar note, I just found [the 'error' topic](https://docs.rs/winnow/latest/winnow/_topic/error/index.html), which shows you how to integrate custom error types. I think it'd be helpful if the tutorial also explained this, how would y'all feel about me describing custom error handling in the Error chapter of the tutorial? Or at least linking to the special topic for error handling.

I feel like custom errors would be beyond the scope of what is intended for the tutorial (getting someone started) though I could see having a "if you'd like to learn more, see the Special Topics: Custom Errors"

> I'd also like to add examples for how to handle errors with custom tokens. The built-in pretty errors when your input type is &str are great. But I think an example for custom tokens, e.g. &[T] would be helpful. How do y'all feel about something like this example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=fc3d358804257748e1f6fe09e5d1e836

This is an interesting one because there are two sides to it
- Connect the dots for people for providing a custom rendering for a `ParseError`
  - this could be an extension on what we already have in https://docs.rs/winnow/latest/winnow/_topic/error/index.html
- Handling of errors for custom tokens (where the right answer depends on what kind of custom token you have)
  - this would fit in with https://docs.rs/winnow/latest/winnow/_topic/stream/index.html

**Comment by epage** [comment]
> I think my earlier playground, or this [expanded playground with custom errors](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ce5ead8928d5997dbd1ca36c5e4b6399) could be a good extension to the existing example. This would address my initial comment, which is "what's the equivalent of [nom::error::convert_error](https://docs.rs/nom/latest/nom/error/fn.convert_error.html) and how do I implement my own".

Except it is fairly complicated for that purpose, including other topics.  I think #104 would more ideally cover this.

> So, a sidebar, but I think that "stream" topic page should explicitly address "I have custom tokens, but I want to use them with Vec/Slice". At least, that's exactly what my parser is doing, and my guess is that's going to be a common starting place -- standard "stream container" type like vec, but custom token type.

Maybe?  There are reasons to have a custom stream besides that, like providing custom behavior for existing token types (e.g. error recovery).

I do agree that it would help to communicate both roles somehow.

**Comment by epage** [comment]
We've expanded our [old error reporting chapter](https://docs.rs/winnow/0.5.40/winnow/_tutorial/chapter_6/index.html) to better talk about context, cuts, and now fancy renderers, see [the new chapter](https://docs.rs/winnow/0.6.20/winnow/_tutorial/chapter_7/index.html).

I think that covers the core of this issue.  There is also some talk about custom tokens though I think that is orthogonal to this with some discussion at https://github.com/winnow-rs/winnow/discussions/591
